WhateverGreen/kern_rad.cpp

```diff
@@ -183,7 +190,96 @@ void RAD::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
 	}
 }
 
+long g_current_brightness_lvl = 0xff7b;
+static void* g_dc_link_ptr = NULL;
+static mach_vm_address_t orig_dce110_set_backlight_level;
+static char wrap_dce110_set_backlight_level(int64_t pipe_ctx, unsigned int a2, int64_t a3) {
+	char ret = FunctionCast(wrap_dce110_set_backlight_level, orig_dce110_set_backlight_level)(pipe_ctx, a2, a3);
+	return ret;
+}
+int64_t amd_set_backlight_lvl_pwm(unsigned int backlight_pwm_u16_16, int64_t ramp) {
+	char ret = 0;
+	if (g_dc_link_ptr) {
+		int64_t v10 = 0;
+		int64_t v9 = *(int64_t *)(*(int64_t *)((int64_t)g_dc_link_ptr + 304) + 944LL);
```

Comment:
These offsets are extracted from the disassembly code of the function _dc_link_dp_set_test_pattern. I checked the code in Linux kernel linux-master/drivers/gpu/drm/amd/display/dc/core/dc_link_dp.c for reference. They may change when Mac OS upgrade, I think it's better to hook _dc_link_set_backlight_level, but it's a __usercall function, I can't determine the value of the argument I can pass into it.

Reply:
It is not a usercall, just IDA failed to determine the type. Correct it by pressing 'Y'.

Reply:
hi vit9696, thank you for your advice, I managed to warp the _dc_link_set_backlight_level call. Now the code is more clean and robust. And I also cleaned out other codes and add a boot option applbkl=3 to enable or disable this patch. I think this code is more compliant with the WEG structure now. I hope it will useful to some else.
However, the cold boot problem still exists. It has beyond my capability.